### PR TITLE
Automated Skyblock update thread detection & posting

### PR DESCRIPTION
**Pull request by wiviw5**
Long wanted automation for sending skyblock update threads in the discord with pings!

### CODE REVIEW

File: src/main/java/net/hypixel/nerdbot/util/watcher/rss/xmlparsers/SkyblockThreadParser.java [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+            }
+        }
+
+
+        @Override
+        public void startDocument() {
+            skyblockForum = new SkyblockForum();
+        }
+
+    }
```
Comment (Aerhhh) [inline src/main/java/net/hypixel/nerdbot/util/watcher/rss/xmlparsers/SkyblockThreadParser.java:161]: Should separate this into its own class

File: src/main/java/net/hypixel/nerdbot/util/watcher/rss/xmlparsers/SkyblockThreadParser.java [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+import javax.xml.parsers.SAXParserFactory;
+import java.io.ByteArrayInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.ArrayList;
+import java.util.List;
+
+@Log4j2
+public class SkyblockThreadParser {
+
```
Comment (Aerhhh) [inline src/main/java/net/hypixel/nerdbot/util/watcher/rss/xmlparsers/SkyblockThreadParser.java:21]: Remove extra empty line

File: src/main/java/net/hypixel/nerdbot/util/watcher/rss/SkyblockUpdateDataHandler.java [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+        MessageCreateBuilder messageCreateBuilder = new MessageCreateBuilder();
+
+        PingableRole role = RoleManager.getPingableRoleByName("SkyBlock Update Alerts");
+
+        if (role != null) {
+            messageCreateBuilder.addContent(RoleManager.formatPingableRoleAsMention(role) + "\n\n");
+        }
+
+        messageCreateBuilder.addContent(hypixelThread.getLink());
+
```
Comment (Aerhhh) [inline src/main/java/net/hypixel/nerdbot/util/watcher/rss/SkyblockUpdateDataHandler.java:35]: Remove empty line

File: src/main/java/net/hypixel/nerdbot/util/watcher/rss/SkyblockUpdateDataHandler.java [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+        if (announcementChannel == null) {
+            log.error("Couldn't find announcement channel!");
+            return;
+        }
+
+
+        MessageCreateBuilder messageCreateBuilder = new MessageCreateBuilder();
+
+        PingableRole role = RoleManager.getPingableRoleByName("SkyBlock Update Alerts");
+
```
Comment (Aerhhh) [inline src/main/java/net/hypixel/nerdbot/util/watcher/rss/SkyblockUpdateDataHandler.java:29]: Remove the empty lines here

**Review by Aerhhh — CHANGES_REQUESTED** [review]

File: src/main/java/net/hypixel/nerdbot/util/watcher/rss/xmlparsers/SkyblockThreadParser.java [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+            }
+        }
+
+        private HypixelThread latestThread() {
+            List<HypixelThread> articleList = skyblockForum.getThreadlist();
+            int latestArticleIndex = articleList.size() - 1;
+            return articleList.get(latestArticleIndex);
+        }
+
+
```
Comment (Aerhhh) [inline src/main/java/net/hypixel/nerdbot/util/watcher/rss/xmlparsers/SkyblockThreadParser.java:145]: Remove extra empty line

File: src/main/java/net/hypixel/nerdbot/util/watcher/rss/xmlparsers/SkyblockThreadParser.java [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+                        log.info("Reading RSS Feed: " + elementValue);
+                        break;
+                    case DESCRIPTION:
+                        log.info("Description: " + elementValue);
+                        break;
+                }
+                return;
+            }
+
+
```
Comment (Aerhhh) [inline src/main/java/net/hypixel/nerdbot/util/watcher/rss/xmlparsers/SkyblockThreadParser.java:119]: Remove extra empty line

**Review by Aerhhh — CHANGES_REQUESTED** [review]

File: src/main/java/net/hypixel/nerdbot/util/watcher/rss/xmlparsers/SkyblockThreadParser.java [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+    public static List<HypixelThread> parseSkyblockThreads(String XML) throws ParserConfigurationException, SAXException, IOException {
+        SAXParserFactory factory = SAXParserFactory.newInstance();
+        SAXParser saxParser = factory.newSAXParser();
+
+        SkyblockThreadHandler skyblockThreadHandler = new SkyblockThreadHandler();
+        InputStream targetStream = new ByteArrayInputStream(XML.getBytes());
+        saxParser.parse(targetStream, skyblockThreadHandler);
+        return skyblockThreadHandler.getSkyblockForum().getThreadlist();
+    }
+
```
Comment (Aerhhh) [inline src/main/java/net/hypixel/nerdbot/util/watcher/rss/xmlparsers/SkyblockThreadParser.java:36]: Remove blank line

File: src/main/java/net/hypixel/nerdbot/util/watcher/rss/xmlparsers/SkyblockThreadParser.java [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.List;
+
+@Log4j2
+public class SkyblockThreadParser {
+
+    public static HypixelThread getLastPostedSkyblockThread(String XML) {
+        try {
+            return parseSkyblockThreads(XML).get(0);
```
Comment (Aerhhh) [inline src/main/java/net/hypixel/nerdbot/util/watcher/rss/xmlparsers/SkyblockThreadParser.java:21]: Also make sure this is changed wherever it's used
```suggestion
    public static HypixelThread getLastPostedSkyBlockThread(String xml) {
        try {
            return parseSkyBlockThreads(xml).get(0);
```

File: src/main/java/net/hypixel/nerdbot/util/watcher/rss/xmlparsers/SkyblockThreadParser.java [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+import javax.xml.parsers.ParserConfigurationException;
+import javax.xml.parsers.SAXParser;
+import javax.xml.parsers.SAXParserFactory;
+import java.io.ByteArrayInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.List;
+
+@Log4j2
+public class SkyblockThreadParser {
```
Comment (Aerhhh) [inline src/main/java/net/hypixel/nerdbot/util/watcher/rss/xmlparsers/SkyblockThreadParser.java:17]:
```suggestion
public class SkyBlockThreadParser {
```

File: src/main/java/net/hypixel/nerdbot/util/watcher/rss/xmlparsers/SkyblockThreadParser.java [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+
+    public static List<HypixelThread> parseSkyblockThreads(String XML) throws ParserConfigurationException, SAXException, IOException {
+        SAXParserFactory factory = SAXParserFactory.newInstance();
+        SAXParser saxParser = factory.newSAXParser();
+
+        SkyblockThreadHandler skyblockThreadHandler = new SkyblockThreadHandler();
+        InputStream targetStream = new ByteArrayInputStream(XML.getBytes());
+        saxParser.parse(targetStream, skyblockThreadHandler);
+        return skyblockThreadHandler.getSkyblockForum().getThreadlist();
+    }
```
Comment (Aerhhh) [inline src/main/java/net/hypixel/nerdbot/util/watcher/rss/xmlparsers/SkyblockThreadParser.java:35]:
```suggestion
    public static List<HypixelThread> parseSkyBlockThreads(String xml) throws ParserConfigurationException, SAXException, IOException {
        SAXParserFactory factory = SAXParserFactory.newInstance();
        SAXParser saxParser = factory.newSAXParser();

        SkyBlockThreadHandler skyBlockThreadHandler = new SkyBlockThreadHandler();
        InputStream targetStream = new ByteArrayInputStream(xml.getBytes());
        saxParser.parse(targetStream, skyBlockThreadHandler);
        return skyBlockThreadHandler.getSkyBlockForum().getThreadList();
    }
```

File: src/main/java/net/hypixel/nerdbot/util/watcher/rss/xmlparsers/SkyblockThreadParser.java [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+
+
+    @Getter
+    @Setter
+    public static class SkyblockForum {
+        private List<HypixelThread> threadlist;
+    }
```
Comment (Aerhhh) [inline src/main/java/net/hypixel/nerdbot/util/watcher/rss/xmlparsers/SkyblockThreadParser.java:42]:
```suggestion
    public static class SkyBlockForum {
        private List<HypixelThread> threadList;
    }
```

File: src/main/java/net/hypixel/nerdbot/util/watcher/rss/xmlparsers/SkyblockThreadHandler.java [hunk 12]
```diff
@@ -0,0 +1,114 @@
...
+            elementValue = new StringBuilder();
+        } else {
+            elementValue.append(ch, start, length);
+        }
+    }
+
+
+    @Override
+    public void startDocument() {
+        skyblockForum = new SkyblockThreadParser.SkyblockForum();
```
Comment (Aerhhh) [inline src/main/java/net/hypixel/nerdbot/util/watcher/rss/xmlparsers/SkyblockThreadHandler.java:111]:
```suggestion
        skyBlockForum = new SkyBlockThreadParser.SkyBlockForum();
```

File: src/main/java/net/hypixel/nerdbot/util/watcher/rss/xmlparsers/SkyblockThreadHandler.java [hunk 13]
```diff
@@ -0,0 +1,114 @@
...
+                latestThread().setGuid(elementValue.toString());
+                break;
+            case CREATOR:
+                latestThread().setCreator(elementValue.toString());
+                break;
+        }
+    }
+
+    private SkyblockThreadParser.HypixelThread latestThread() {
+        List<SkyblockThreadParser.HypixelThread> articleList = skyblockForum.getThreadlist();
```
Comment (Aerhhh) [inline src/main/java/net/hypixel/nerdbot/util/watcher/rss/xmlparsers/SkyblockThreadHandler.java:93]:
```suggestion
    private SkyBlockThreadParser.HypixelThread latestThread() {
        List<SkyBlockThreadParser.HypixelThread> articleList = skyBlockForum.getThreadList();
```

File: src/main/java/net/hypixel/nerdbot/util/watcher/rss/xmlparsers/SkyblockThreadHandler.java [hunk 14]
```diff
@@ -0,0 +1,114 @@
...
+                    finishedReadingPrelude = true;
+                    break;
+            }
+            return;
+        }
+
+        // Now that we've completed every starter tag, we can move onto the actual threads.
+        switch (qName) {
+            case ITEM:
+                skyblockForum.getThreadlist().add(new SkyblockThreadParser.HypixelThread());
```
Comment (Aerhhh) [inline src/main/java/net/hypixel/nerdbot/util/watcher/rss/xmlparsers/SkyblockThreadHandler.java:50]:
```suggestion
                skyBlockForum.getThreadList().add(new SkyBlockThreadParser.HypixelThread());
```

File: src/main/java/net/hypixel/nerdbot/util/watcher/rss/xmlparsers/SkyblockThreadHandler.java [hunk 15]
```diff
@@ -0,0 +1,114 @@
...
+    private static final String GUID = "guid";
+    private static final String CREATOR = "dc:creator";
+
+    @Override
+    public void startElement(String uri, String localName, String qName, Attributes attributes) {
+        // We first handle the prelude tags.
+        if (!finishedReadingPrelude) {
+            switch (qName) {
+                case CHANNEL:
+                    skyblockForum.setThreadlist(new ArrayList<>());
```
Comment (Aerhhh) [inline src/main/java/net/hypixel/nerdbot/util/watcher/rss/xmlparsers/SkyblockThreadHandler.java:34]:
```suggestion
                    skyBlockForum.setThreadList(new ArrayList<>());
```

File: src/main/java/net/hypixel/nerdbot/util/watcher/rss/xmlparsers/SkyblockThreadHandler.java [hunk 16]
```diff
@@ -0,0 +1,114 @@
...
+
+import java.util.ArrayList;
+import java.util.List;
+
+@Log4j2
+public class SkyblockThreadHandler extends DefaultHandler {
+
+    private static boolean finishedReadingPrelude = false;
+    @Getter
+    private SkyblockThreadParser.SkyblockForum skyblockForum;
```
Comment (Aerhhh) [inline src/main/java/net/hypixel/nerdbot/util/watcher/rss/xmlparsers/SkyblockThreadHandler.java:16]:
```suggestion
    private SkyBlockThreadParser.SkyBlockForum skyBlockForum;
```

File: src/main/java/net/hypixel/nerdbot/util/watcher/rss/xmlparsers/SkyblockThreadHandler.java [hunk 17]
```diff
@@ -0,0 +1,114 @@
...
+import lombok.Getter;
+import lombok.extern.log4j.Log4j2;
+import org.xml.sax.Attributes;
+import org.xml.sax.helpers.DefaultHandler;
+
+import java.util.ArrayList;
+import java.util.List;
+
+@Log4j2
+public class SkyblockThreadHandler extends DefaultHandler {
```
Comment (Aerhhh) [inline src/main/java/net/hypixel/nerdbot/util/watcher/rss/xmlparsers/SkyblockThreadHandler.java:12]:
```suggestion
public class SkyBlockThreadHandler extends DefaultHandler {
```

File: src/main/java/net/hypixel/nerdbot/util/watcher/rss/SkyblockUpdateDataHandler.java [hunk 18]
```diff
@@ -0,0 +1,34 @@
...
+import net.dv8tion.jda.api.utils.messages.MessageCreateBuilder;
+import net.hypixel.nerdbot.NerdBotApp;
+import net.hypixel.nerdbot.bot.config.ChannelConfig;
+import net.hypixel.nerdbot.channel.ChannelManager;
+import net.hypixel.nerdbot.role.PingableRole;
+import net.hypixel.nerdbot.role.RoleManager;
+import net.hypixel.nerdbot.util.watcher.rss.xmlparsers.SkyblockThreadParser.HypixelThread;
+
+@Log4j2
+public class SkyblockUpdateDataHandler {
```
Comment (Aerhhh) [inline src/main/java/net/hypixel/nerdbot/util/watcher/rss/SkyblockUpdateDataHandler.java:14]:
```suggestion
public class SkyBlockUpdateDataHandler {
```

File: src/main/java/net/hypixel/nerdbot/util/watcher/rss/HypixelThreadURLWatcher.java [hunk 19]
```diff
@@ -0,0 +1,57 @@
...
+        active = true;
+    }
+
+    @Override
+    public void watchOnce(URLWatcher.DataHandler handler) {
+        String newContent = fetchContent();
+        active = true;
+
+        if (newContent != null && !newContent.equals(getLastContent())) {
+            handleThread(getLastPostedSkyblockThread(newContent));
```
Comment (Aerhhh) [inline src/main/java/net/hypixel/nerdbot/util/watcher/rss/HypixelThreadURLWatcher.java:50]:
```suggestion
            handleThread(getLastPostedSkyBlockThread(newContent));
```

File: src/main/java/net/hypixel/nerdbot/util/watcher/rss/HypixelThreadURLWatcher.java [hunk 20]
```diff
@@ -0,0 +1,57 @@
...
+    }
+
+    @Override
+    public void startWatching(long interval, TimeUnit unit, URLWatcher.DataHandler handler) {
+        timer.scheduleAtFixedRate(new TimerTask() {
+            @Override
+            public void run() {
+                String newContent = fetchContent();
+                if (newContent != null && !newContent.equals(lastContent)) {
+                    handleThread(getLastPostedSkyblockThread(newContent));
```
Comment (Aerhhh) [inline src/main/java/net/hypixel/nerdbot/util/watcher/rss/HypixelThreadURLWatcher.java:33]:
```suggestion
                    handleThread(getLastPostedSkyBlockThread(newContent));
```

File: src/main/java/net/hypixel/nerdbot/util/watcher/rss/xmlparsers/SkyblockThreadHandler.java [hunk 21]
```diff
@@ -0,0 +1,114 @@
...
+    public void endElement(String uri, String localName, String qName) {
+        // We first read the prelude tags, once complete we can move onto the rest of the tags.
+        if (!finishedReadingPrelude) {
+            switch (qName) {
+                case TITLE:
+                    log.info("Reading RSS Feed: " + elementValue);
+                    break;
+                case DESCRIPTION:
+                    log.info("Description: " + elementValue);
+                    break;
```
Comment (Aerhhh) [inline src/main/java/net/hypixel/nerdbot/util/watcher/rss/xmlparsers/SkyblockThreadHandler.java:68]: I would probably change these to debug instead

File: src/main/java/net/hypixel/nerdbot/util/watcher/rss/SkyblockUpdateDataHandler.java [hunk 22]
```diff
@@ -0,0 +1,34 @@
...
+
+    public static void handleThread(HypixelThread hypixelThread) {
+        ChannelConfig config = NerdBotApp.getBot().getConfig().getChannelConfig();
+        TextChannel announcementChannel = ChannelManager.getChannel(config.getAnnouncementChannelId());
+
```
Comment (Aerhhh) [inline src/main/java/net/hypixel/nerdbot/util/watcher/rss/SkyblockUpdateDataHandler.java:19]:
```suggestion
```

**Review by Aerhhh — CHANGES_REQUESTED** [review]

**Comment by wiviw5** [comment]
Known Issues:

1: articleList being set to null for some reason, possibly due to the code running at the same time

2: For some reason the new content changes when there are no changes.

These two will need to be solved + possibly more.

For 1 thinking we can maybe change up how the list of threads for output is stored?
(Potentially including a list all the way back at the start of the runtime methods to store their results in instead of in class?)

For 2, I am thinking downloading the XML everytime leads to something changing between downloads, I think the proper solution is possibly to rely on something else? Perhaps parsing each minute the new threads and seeing whats new?

File: src/main/java/net/hypixel/nerdbot/util/watcher/rss/HypixelThreadURLWatcher.java [hunk 23]
```diff
@@ -0,0 +1,121 @@
...
+    }
+
+    public int getHighestGuid(){
+        List<HypixelThread> hypixelThreads = fetchAndParseContent();
+        int[] guidList = new int[hypixelThreads.size()];
+        for (int x = 0; x < hypixelThreads.size(); x++) {
+            guidList[x] = hypixelThreads.get(x).getGuid();
+        }
+        return Arrays.stream(guidList).max().getAsInt();
+    }
```
Comment (Aerhhh) [inline src/main/java/net/hypixel/nerdbot/util/watcher/rss/HypixelThreadURLWatcher.java:91]:
```suggestion
    public int getHighestGuid() {
        List<HypixelThread> hypixelThreads = fetchAndParseContent();
        int[] guidList = new int[hypixelThreads.size()];
        
        for (int x = 0; x < hypixelThreads.size(); x++) {
            guidList[x] = hypixelThreads.get(x).getGuid();
        }
        
        return Arrays.stream(guidList).max().getAsInt();
    }
```

File: src/main/java/net/hypixel/nerdbot/util/watcher/rss/HypixelThreadURLWatcher.java [hunk 24]
```diff
@@ -0,0 +1,121 @@
...
+        if (hypixelThreads == null) {
+            return;
+        }
+        for (HypixelThread thread : hypixelThreads) {
+            if (thread.getGuid() > lastGuid) {
+                log.debug("Watched " + url + " and found newest Guid!\nOld Guid: " + lastGuid + "\nNew Guid: " + thread.getGuid());
+                lastGuid = thread.getGuid();
+                handleThread(thread);
+            }
+        }
```
Comment (Aerhhh) [inline src/main/java/net/hypixel/nerdbot/util/watcher/rss/HypixelThreadURLWatcher.java:79]:
```suggestion
        if (hypixelThreads == null) {
            return;
        }
        
        for (HypixelThread thread : hypixelThreads) {
            if (thread.getGuid() > lastGuid) {
                log.debug("Watched " + url + " and found newest Guid!\nOld Guid: " + lastGuid + "\nNew Guid: " + thread.getGuid());
                lastGuid = thread.getGuid();
                handleThread(thread);
            }
        }
```

File: src/main/java/net/hypixel/nerdbot/util/watcher/rss/HypixelThreadURLWatcher.java [hunk 25]
```diff
@@ -0,0 +1,121 @@
...
+                    return;
+                }
+                for (HypixelThread thread : hypixelThreads) {
+                    if (thread.getGuid() > lastGuid) {
+                        log.debug("Watched " + url + " and found newest Guid!\nOld Guid: " + lastGuid + "\nNew Guid: " + thread.getGuid());
+                        lastGuid = thread.getGuid();
+                        handleThread(thread);
+                    }
+                }
+            }
```
Comment (Aerhhh) [inline src/main/java/net/hypixel/nerdbot/util/watcher/rss/HypixelThreadURLWatcher.java:59]:
```suggestion
            public void run() {
                List<HypixelThread> hypixelThreads = fetchAndParseContent();
                if (hypixelThreads == null) {
                    return;
                }
                
                for (HypixelThread thread : hypixelThreads) {
                    if (thread.getGuid() > lastGuid) {
                        log.debug("Watched " + url + " and found newest Guid!\nOld GUID: " + lastGuid + "\nNew GUID: " + thread.getGuid());
                        lastGuid = thread.getGuid();
                        handleThread(thread);
                    }
                }
            }
```

File: src/main/java/net/hypixel/nerdbot/util/watcher/rss/HypixelThreadURLWatcher.java [hunk 26]
```diff
@@ -0,0 +1,121 @@
...
+import okhttp3.OkHttpClient;
+import okhttp3.Request;
+import okhttp3.Response;
+
+import java.io.IOException;
+import java.util.*;
+import java.util.concurrent.TimeUnit;
+
+import static net.hypixel.nerdbot.util.watcher.rss.SkyBlockUpdateDataHandler.handleThread;
+import static net.hypixel.nerdbot.util.watcher.rss.xmlparsers.SkyBlockThreadParser.parseSkyBlockThreads;
```
Comment (Aerhhh) [inline src/main/java/net/hypixel/nerdbot/util/watcher/rss/HypixelThreadURLWatcher.java:16]: Remove these

File: src/main/java/net/hypixel/nerdbot/util/watcher/rss/SkyBlockUpdateDataHandler.java [hunk 27]
```diff
@@ -0,0 +1,39 @@
...
+
+        MessageCreateBuilder messageCreateBuilder = new MessageCreateBuilder();
+        PingableRole role = RoleManager.getPingableRoleByName("SkyBlock Update Alerts");
+        if (role != null) {
+            messageCreateBuilder.addContent(RoleManager.formatPingableRoleAsMention(role) + "\n\n");
+        }
+
+        messageCreateBuilder.addContent(hypixelThread.getLink());
+        announcementChannel.sendMessage(messageCreateBuilder.build()).queue();
+    }
```
Comment (Aerhhh) [inline src/main/java/net/hypixel/nerdbot/util/watcher/rss/SkyBlockUpdateDataHandler.java:38]:
```suggestion
    public static void handleThread(HypixelThread hypixelThread) {
        ChannelConfig config = NerdBotApp.getBot().getConfig().getChannelConfig();
        TextChannel announcementChannel = ChannelManager.getChannel(config.getAnnouncementChannelId());
        if (announcementChannel == null) {
            log.error("Couldn't find announcement channel!");
            return;
        }
        
        // Simple Check to make sure only SkyBlock threads are sent.
        if (!hypixelThread.getForum().equals("SkyBlock Patch Notes")){
            if (!hypixelThread.getTitle().contains("SkyBlock")){
                return;
            }
        }

        MessageCreateBuilder messageCreateBuilder = new MessageCreateBuilder();
        PingableRole role = RoleManager.getPingableRoleByName("SkyBlock Update Alerts");
        if (role != null) {
            messageCreateBuilder.addContent(RoleManager.formatPingableRoleAsMention(role) + "\n\n");
        }

        messageCreateBuilder.addContent(hypixelThread.getLink());
        announcementChannel.sendMessage(messageCreateBuilder.build()).queue();
    }
```

File: src/main/java/net/hypixel/nerdbot/util/watcher/rss/xmlparsers/SkyBlockThreadHandler.java [hunk 28]
```diff
@@ -0,0 +1,117 @@
...
+    private static String forum = null;
+    private static final String DESCRIPTION = "description";
+    private static final String ATOMLINK = "atom:link";
+    private static final String CHANNEL = "channel";
+    private static final String ITEM = "item";
+    private static final String TITLE = "title";
+    private static final String PUBDATE = "pubDate";
+    private static final String LINK = "link";
+    private static final String GUID = "guid";
+    private static final String CREATOR = "dc:creator";
```
Comment (Aerhhh) [inline src/main/java/net/hypixel/nerdbot/util/watcher/rss/xmlparsers/SkyBlockThreadHandler.java:26]:
```suggestion
public class SkyBlockThreadHandler extends DefaultHandler {

    private static final String DESCRIPTION = "description";
    private static final String ATOMLINK = "atom:link";
    private static final String CHANNEL = "channel";
    private static final String ITEM = "item";
    private static final String TITLE = "title";
    private static final String PUBDATE = "pubDate";
    private static final String LINK = "link";
    private static final String GUID = "guid";
    private static final String CREATOR = "dc:creator";

    @Getter
    private SkyBlockThreadParser.SkyBlockForum skyBlockForum;
    private StringBuilder elementValue;
    private static String forum = null;
```

File: src/main/java/net/hypixel/nerdbot/util/watcher/rss/xmlparsers/SkyBlockThreadHandler.java [hunk 29]
```diff
@@ -0,0 +1,117 @@
...
+                break;
+        }
+    }
+
+    private SkyBlockThreadParser.HypixelThread latestThread() {
+        List<SkyBlockThreadParser.HypixelThread> threadList = skyBlockForum.getThreadList();
+        int latestArticleIndex = threadList.size() - 1;
+        return threadList.get(latestArticleIndex);
+    }
+
```
Comment (Aerhhh) [inline src/main/java/net/hypixel/nerdbot/util/watcher/rss/xmlparsers/SkyBlockThreadHandler.java:100]:
```suggestion
```

File: src/main/java/net/hypixel/nerdbot/util/watcher/rss/xmlparsers/SkyBlockThreadHandler.java [hunk 30]
```diff
@@ -0,0 +1,117 @@
...
+    @Override
+    public void characters(char[] ch, int start, int length) {
+        if (elementValue == null) {
+            elementValue = new StringBuilder();
+        } else {
+            elementValue.append(ch, start, length);
+        }
+    }
+
```
Comment (Aerhhh) [inline src/main/java/net/hypixel/nerdbot/util/watcher/rss/xmlparsers/SkyBlockThreadHandler.java:110]:
```suggestion
```

File: src/main/java/net/hypixel/nerdbot/util/watcher/rss/xmlparsers/SkyBlockThreadHandler.java [hunk 31]
```diff
@@ -0,0 +1,117 @@
...
+                    break;
+                case TITLE:
+                    elementValue = new StringBuilder();
+                    break;
+                case ATOMLINK:
+                    // This tag should mark the end of prelude, and thus we stop checking for these here.
+                    skyBlockForum.setFinishedReadingPrelude(true);
+                    break;
+            }
+            return;
```
Comment (Aerhhh) [inline src/main/java/net/hypixel/nerdbot/util/watcher/rss/xmlparsers/SkyBlockThreadHandler.java:44]:
```suggestion
            }
            
            return;
```

File: src/main/java/net/hypixel/nerdbot/util/watcher/rss/xmlparsers/SkyBlockThreadHandler.java [hunk 32]
```diff
@@ -0,0 +1,117 @@
...
+        if (!skyBlockForum.isFinishedReadingPrelude()) {
+            switch (qName) {
+                case TITLE:
+                    forum = elementValue.toString();
+                    break;
+                case DESCRIPTION:
+                    // For future use if someone wants to know the description of this forum.
+                    break;
+            }
+            return;
```
Comment (Aerhhh) [inline src/main/java/net/hypixel/nerdbot/util/watcher/rss/xmlparsers/SkyBlockThreadHandler.java:70]:
```suggestion
            }
            
            return;
```

**Review by Aerhhh — CHANGES_REQUESTED** [review]

**Pull request merged**